# Ecommerce Sales Dashboard

## Notebook 04 – SQL Analysis

### Author
Divya Sesilia

### Project
End-to-End Data Analytics Project using SQL, Python, Excel, and Power BI

### Dataset
Superstore Sales Dataset (Cleaned)

### Database
SQLite

### Tools Used

- SQL
- SQLite
- Pandas
- Jupyter Notebook

### Notebook Version

1.0

---

## Objective

The objective of this notebook is to answer key business questions using SQL queries on the cleaned dataset.

---

## Business Questions

- What are the total sales and profit?
- Which categories generate the highest sales?
- Which sub-categories are most profitable?
- Which regions perform the best?
- Who are the top customers?
- How do discounts affect profit?
- What are the monthly sales trends?

## Import Required Libraries

Import the libraries required to connect Python with SQLite and execute SQL queries.

In [162]:
import pandas as pd
import sqlite3

## Load the Cleaned Dataset

Load the cleaned dataset created in Notebook 02.

In [163]:
df = pd.read_csv("../data/cleaned/superstore_cleaned.csv")

print(f"Dataset Shape: {df.shape}")

df.head()

Dataset Shape: (9994, 24)


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit,Order Year,Order Month,Order Quarter
0,1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136,2016,November,4
1,2,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820,2016,November,4
2,3,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714,2016,June,2
3,4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310,2015,October,4
4,5,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164,2015,October,4


## Create SQLite Database

Create an in-memory SQLite database and load the cleaned dataset into a SQL table.

In [164]:
conn = sqlite3.connect(":memory:")

df.to_sql(
    "superstore",
    conn,
    index=False,
    if_exists="replace"
)

print("SQLite database created successfully.")
print("Table 'superstore' is ready for SQL analysis.")

SQLite database created successfully.
Table 'superstore' is ready for SQL analysis.


## SQL Table Preview

Display the first five records from the SQLite table to verify that the data has been loaded correctly.

In [165]:
query = """
SELECT *
FROM superstore
LIMIT 5;
"""

pd.read_sql(query, conn)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit,Order Year,Order Month,Order Quarter
0,1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136,2016,November,4
1,2,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820,2016,November,4
2,3,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714,2016,June,2
3,4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310,2015,October,4
4,5,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164,2015,October,4


# Data Validation

Before performing SQL analysis, verify the dataset structure and quality.

In [166]:
query = """
SELECT
    COUNT(*) AS Total_Rows,
    ROUND(SUM(Sales),2) AS Total_Sales
FROM superstore;
"""

pd.read_sql(query, conn)

,Total_Rows,Total_Sales
0,9994,2297200.86


## Validation Result

- The cleaned dataset has been successfully loaded into the SQLite database.
- The total number of records matches the cleaned dataset.
- The SQL table is ready for business analysis.


# Sales Analysis

Sales analysis measures the company's overall revenue performance using SQL queries.

## Business Question

What are the total sales generated by the company?

In [167]:
query = """
SELECT
ROUND(SUM(Sales),2) AS Total_Sales
FROM superstore;
"""

pd.read_sql(query, conn)

,Total_Sales
0,2297200.86


## Business Insight

The company generated approximately **$2.30 million** in total sales.

This represents the total revenue generated from all customer transactions.

## Business Recommendation

Management should monitor total sales regularly to evaluate business growth and measure the effectiveness of sales strategies.

## Business Question

What is the average sales value per transaction?

In [168]:
query = """
SELECT
ROUND(AVG(Sales),2) AS Average_Sales
FROM superstore;
"""

pd.read_sql(query, conn)

,Average_Sales
0,229.86


## Business Insight

The average sales value per transaction is approximately **$229.86**.

This metric helps identify changes in customer purchasing behavior.

## Business Recommendation

Increasing average order value through product bundling and upselling can improve revenue.

# Profit Analysis

Profit analysis evaluates the company's financial performance using SQL queries.

## Business Question

What is the total profit generated by the company?

In [169]:
query = """
SELECT
ROUND(SUM(Profit),2) AS Total_Profit
FROM superstore;
"""

pd.read_sql(query, conn)

,Total_Profit
0,286397.02


## Business Insight

The company generated approximately **$286,397** in total profit.

This indicates that the business is profitable overall.

## Business Recommendation

Focus on improving profit margins while maintaining sales growth.

## Business Question

What is the average profit earned per transaction?

In [170]:
query = """
SELECT
    ROUND(AVG(Profit),2) AS Average_Profit
FROM superstore;
"""

pd.read_sql(query, conn)

,Average_Profit
0,28.66


## Business Insight

The average profit earned per transaction is approximately **$28.66**.

This metric helps evaluate the profitability of individual customer transactions.

## Business Recommendation

Increase average profit by promoting higher-margin products and reducing excessive discounting.

# Category Analysis

Category analysis compares the sales and profitability of major product categories.

This helps identify which categories contribute the most to the company's overall business performance.

## Business Question

Which product category generates the highest sales?

In [171]:
query = """
SELECT
    Category,
    ROUND(SUM(Sales),2) AS Total_Sales
FROM superstore
GROUP BY Category
ORDER BY Total_Sales DESC;
"""

pd.read_sql(query, conn)

,Category,Total_Sales
0,Technology,836154.03
1,Furniture,741999.80
2,Office Supplies,719047.03


## Business Insight

Technology generates the highest total sales, followed by Furniture and Office Supplies.

This indicates that Technology products are the primary driver of the company's revenue.

## Business Recommendation

Continue investing in Technology products while identifying opportunities to improve sales in the Furniture and Office Supplies categories.

## Business Question

Which product category generates the highest profit?

In [172]:
query = """
SELECT
    Category,
    ROUND(SUM(Profit),2) AS Total_Profit
FROM superstore
GROUP BY Category
ORDER BY Total_Profit DESC;
"""

pd.read_sql(query, conn)

,Category,Total_Profit
0,Technology,145454.95
1,Office Supplies,122490.80
2,Furniture,18451.27


## Business Insight

Technology is the most profitable category, while Furniture generates comparatively lower profit despite strong sales.

This suggests that high sales do not always translate into high profitability.

## Business Recommendation

Improve profitability in the Furniture category by reviewing pricing strategies, procurement costs, and discount policies.

# Sub-Category Analysis

Sub-category analysis provides a detailed view of product performance within each category.

It helps identify the highest-performing and lowest-performing product groups.

## Business Question

Which sub-categories generate the highest sales?

In [173]:
query = """
SELECT
    [Sub-Category],
    ROUND(SUM(Sales),2) AS Total_Sales
FROM superstore
GROUP BY [Sub-Category]
ORDER BY Total_Sales DESC;
"""

pd.read_sql(query, conn)

,Sub-Category,Total_Sales
0,Phones,330007.05
1,Chairs,328449.10
2,Storage,223843.61
3,Tables,206965.53
4,Binders,203412.73
5,Machines,189238.63
6,Accessories,167380.32
7,Copiers,149528.03
8,Bookcases,114880.00
9,Appliances,107532.16


## Business Insight

Phones, Chairs, and Storage are among the highest-selling sub-categories.

These products contribute a significant share of the company's overall revenue.

## Business Recommendation

Maintain sufficient inventory levels and prioritize marketing efforts for high-selling sub-categories to sustain revenue growth.

## Business Question

Which sub-categories generate the highest profit?

In [174]:
query = """
SELECT
    [Sub-Category],
    ROUND(SUM(Profit),2) AS Total_Profit
FROM superstore
GROUP BY [Sub-Category]
ORDER BY Total_Profit DESC;
"""

pd.read_sql(query, conn)

,Sub-Category,Total_Profit
0,Copiers,55617.82
1,Phones,44515.73
2,Accessories,41936.64
3,Paper,34053.57
4,Binders,30221.76
5,Chairs,26590.17
6,Storage,21278.83
7,Appliances,18138.01
8,Furnishings,13059.14
9,Envelopes,6964.18


## Business Insight

Copiers generate the highest profit, while Tables and Bookcases produce comparatively lower profits.

Although Tables and Bookcases contribute to sales, their profitability is significantly lower than other sub-categories, indicating potential issues with pricing, discounts, or operational costs.

## Business Recommendation

Review pricing strategies, supplier costs, and discount policies for Tables and Bookcases to improve profitability while maintaining sales performance.

# Region Analysis

Region analysis evaluates sales and profit performance across different geographical regions.

Understanding regional performance helps identify high-performing markets and areas that require business improvement.

## Business Question

Which region generates the highest sales?

In [175]:
query = """
SELECT
    Region,
    ROUND(SUM(Sales),2) AS Total_Sales
FROM superstore
GROUP BY Region
ORDER BY Total_Sales DESC;
"""

pd.read_sql(query, conn)

,Region,Total_Sales
0,West,725457.82
1,East,678781.24
2,Central,501239.89
3,South,391721.91


## Business Insight

The West region generates the highest total sales, followed by the East region.

This indicates that these regions contribute significantly to the company's overall revenue.

## Business Recommendation

Continue investing in high-performing regions while developing targeted marketing and sales strategies to improve performance in lower-performing regions.

## Business Question

Which region generates the highest profit?

In [176]:
query = """
SELECT
    Region,
    ROUND(SUM(Profit),2) AS Total_Profit
FROM superstore
GROUP BY Region
ORDER BY Total_Profit DESC;
"""

pd.read_sql(query, conn)

,Region,Total_Profit
0,West,108418.45
1,East,91522.78
2,South,46749.43
3,Central,39706.36


## Business Insight

The West region contributes the highest profit, while some regions generate comparatively lower profits.

Comparing sales and profit reveals that higher sales do not always guarantee higher profitability.

## Business Recommendation

Analyze the operational practices of high-performing regions and implement similar strategies in lower-performing regions to improve profitability.

# Customer Analysis

Customer analysis identifies the company's highest-value customers based on sales.

Understanding customer contribution supports customer retention and loyalty initiatives.

## Business Question

Who are the top 10 customers by total sales?

In [177]:
query = """
SELECT
    [Customer Name],
    ROUND(SUM(Sales),2) AS Total_Sales,
    ROUND(SUM(Profit),2) AS Total_Profit
FROM superstore
GROUP BY [Customer Name]
ORDER BY Total_Sales DESC
LIMIT 10;
"""

pd.read_sql(query, conn)

,Customer Name,Total_Sales,Total_Profit
0,Sean Miller,25043.05,-1980.74
1,Tamara Chand,19052.22,8981.32
2,Raymond Buch,15117.34,6976.10
3,Tom Ashbrook,14595.62,4703.79
4,Adrian Barton,14473.57,5444.81
5,Ken Lonsdale,14175.23,806.85
6,Sanjit Chand,14142.33,5757.41
7,Hunter Lopez,12873.30,5622.43
8,Sanjit Engle,12209.44,2650.68
9,Christopher Conant,12129.07,2177.05


## Business Insight

A small group of customers contributes a significant portion of total sales.

These customers represent valuable long-term business relationships.

## Business Recommendation

Strengthen customer loyalty through personalized services, targeted promotions, and retention programs for high-value customers.

# Discount Analysis

Discount analysis examines how pricing discounts influence business profitability.

Understanding this relationship helps optimize promotional strategies.

## Business Question

How does discount affect average profit?

In [178]:
query = """
SELECT
    Discount,
    ROUND(AVG(Profit),2) AS Average_Profit
FROM superstore
GROUP BY Discount
ORDER BY Discount;
"""

pd.read_sql(query, conn)

,Discount,Average_Profit
0,0.00,66.90
1,0.10,96.06
2,0.15,27.29
3,0.20,24.70
4,0.30,-45.68
5,0.32,-88.56
6,0.40,-111.93
7,0.45,-226.65
8,0.50,-310.70
9,0.60,-43.08


## Business Insight

Higher discount levels generally result in lower average profit.

Several heavily discounted transactions generate losses despite contributing to sales.

## Business Recommendation

Review discount policies carefully to balance increased sales volume with sustainable profitability.

# Monthly Sales Trend

Monthly sales trend analysis identifies seasonal purchasing behavior throughout the year.

This helps businesses plan inventory and promotional campaigns.

## Business Question

What are the monthly sales trends?

In [179]:
query = """
SELECT
    [Order Year],
    [Order Month],
    ROUND(SUM(Sales),2) AS Total_Sales
FROM superstore
GROUP BY
    [Order Year],
    [Order Month]
ORDER BY
    [Order Year],
    MIN([Order Date]);
"""

pd.read_sql(query, conn)

,Order Year,Order Month,Total_Sales
0,2014,January,14236.90
1,2014,February,4519.89
2,2014,March,55691.01
3,2014,April,28295.35
4,2014,May,23648.29
5,2014,June,34595.13
6,2014,July,33946.39
7,2014,August,27909.47
8,2014,September,81777.35
9,2014,October,31453.39


## Business Insight

Sales fluctuate across different months, indicating seasonal demand patterns.

Certain months consistently generate higher revenue than others.

## Business Recommendation

Plan marketing campaigns, inventory replenishment, and staffing based on expected seasonal demand.

# Yearly Sales Trend

Yearly sales trend analysis evaluates long-term business growth.

Monitoring annual performance supports strategic planning and forecasting.

## Business Question

How have sales changed over the years?

In [180]:
query = """
SELECT
    [Order Year],
    ROUND(SUM(Sales),2) AS Total_Sales
FROM superstore
GROUP BY
    [Order Year]
ORDER BY
    [Order Year];
"""

pd.read_sql(query, conn)

,Order Year,Total_Sales
0,2014,484247.50
1,2015,470532.51
2,2016,609205.60
3,2017,733215.26


## Business Insight

The company demonstrates an overall upward trend in annual sales, indicating positive business growth over time.

Monitoring yearly performance helps assess long-term business success.

## Business Recommendation

Continue monitoring annual sales trends to support strategic planning, budgeting, and future business expansion.

---

# Executive Summary

## Key Findings

- Total sales exceeded **$2.30 million**.
- Total profit exceeded **$286 thousand**.
- Technology generated the highest sales and profit among product categories.
- Copiers were the most profitable sub-category, while Tables and Bookcases showed comparatively lower profitability.
- The West region generated the highest sales and profit.
- A small number of customers contributed significantly to overall revenue.
- Higher discounts were generally associated with lower profitability.
- Sales exhibited seasonal patterns and overall year-over-year growth.

## Business Recommendations

- Continue investing in high-performing categories and regions.
- Optimize pricing and discount strategies for low-profit products.
- Strengthen relationships with high-value customers.
- Use seasonal trends for inventory planning and promotional campaigns.
- Continuously monitor key business metrics to support data-driven decision-making.

## Conclusion

This SQL analysis validates the insights obtained through Python-based exploratory analysis and demonstrates how SQL can be used to answer key business questions efficiently. The results provide a strong foundation for building interactive dashboards in Excel and Power BI.

## Close Database Connection

Close the SQLite connection to release system resources.

In [ ]:
conn.close()

print("SQLite database closed successfully.")
print("Table 'superstore' is ready for SQL analysis.")

SQLite database created successfully.
Table 'superstore' is ready for SQL analysis.


# Notebook Completed

The SQL analysis has been successfully completed.

The insights generated in this notebook will be used to build interactive dashboards in Excel and Power BI in the next stages of the project.